pkg/config/decode.go

```diff
@@ -0,0 +1,113 @@
+package config
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+
+	"github.com/pelletier/go-toml"
+)
+
+// decode parses a TOML tree (and environment variables) according to field tags set in the Config struct.
+//
+// Essentially, it expects `toml` and `env` field tags to be set. Cf. Config for reference.
+func decode(tree *toml.Tree) (conf Config, err error) {
```

Comment:
it would be nice if this was a pure function instead of embedding a dependency on calls to the `os` package to resolve environment variables. Some suggestions for alternative approaches:

1. Consider adding an `env []string` parameter to this func (and to `decodeContext`) which a caller MAY populate with the results of `https://golang.org/pkg/os/#Environ`, or else;
2. Consider adding a `lookupEnv func(string)(string,bool)` parameter to this func (and to `decodeContext`) which a caller may populate with a reference to `os.LookupEnv`.

I tend to prefer (2). This would also simplify unit testing: you wouldn't need to muck around with setting/restoring the process environment just to unit test the decoding logic here.

Reply:
Indeed that makes unit tests easier, 1. might have been good as it makes only one syscall, but I went for the second which is also nice.